### PR TITLE
[RAPPS][APPWIZ] Allow to open "Installed" RAPPS section via command-line

### PR DESCRIPTION
**Pull request by HBelusca**
## Purpose

Open the "Installed" RAPPS section when trying to view the already-installed programs from appwiz.cpl.
On behalf of @binarymaster .

JIRA issue: [CORE-18981](https://jira.reactos.org/browse/CORE-18981)

## Proposed changes

```
[RAPPS] Add /APPWIZ command-line key to open "Installed" section
Addendum to ab7ddc44.
```
```
[RAPPS] Open "Installed" section in Appwiz-mode when RAPPS already runs
```
```
[APPWIZ] Use new Appwiz-mode to start RAPPS with "Installed" section open
```

### CODE REVIEW

**Comment by JoachimHenze** [comment]
Can we make rapps not download all those txt files also when it's just started in the appwiz-mode?
It could try to do that e.g. when explicitly requested or when user clicks on the "available" thing for the first time.
Possible @HBelusca ?

**Comment by HBelusca** [comment]
> Can we make rapps not download all those txt files also when it's just started in the appwiz-mode? It could try to do that e.g. when explicitly requested or when user clicks on the "available" thing for the first time. Possible @HBelusca ?

```
BOT Stas'M:
it's a good idea, but it has to be done separately :)

BOT Stas'M:
I think I can even do it, ask Joachim to file a new ticket, I&#x27;ll assign myself to it
(but need appwiz stuff mergefested first)
```

File: base/applications/rapps/gui.cpp [hunk 1]
```diff
@@ -557,6 +557,11 @@ CMainWindow::OnCommand(WPARAM wParam, LPARAM lParam)
             case ID_CHECK_ALL:
                 m_ApplicationView->CheckAll();
                 break;
+
+            case ID_ACTIVATE_APPWIZ:
```
Comment (HBelusca) [inline base/applications/rapps/gui.cpp:561]: Why is this needed even with the initial `SelectItem` call to the correct tree node in the `InitCategoriesList` function above?

**Review by HBelusca — COMMENTED** [review]
(I approve the PR but as I created it I cannot push the button.)

File: base/applications/rapps/include/appdb.h [hunk 2]
```diff
@@ -18,6 +18,8 @@ class CAppDB
   public:
     CAppDB(const CStringW &path);
 
+    BOOL bAppwizMode;
```
Comment (learn-more) [inline base/applications/rapps/include/appdb.h:21]: How is this at all related to the application db?

Comment (HBelusca) [reply base/applications/rapps/include/appdb.h:21]: Stas'M:

>   How is this at all related to the application db?

I didn&#x27;t find a namespace to share a variable between unattended.cpp and gui.cpp -- it&#x27;s not related to the appdb, but it&#x27;s the simplest way to add it

Another options are:
● Add a global variable somewhere
● Add field to Settings.xxx global variable

In the first case I&#x27;m not fan of adding new global vars :) Second case would have required some additional logic and hacks.

**Comment by HBelusca** [comment]
> > > Can we make rapps not download all those txt files also when it's just started in the appwiz-mode? It could try to do that e.g. when explicitly requested or when user clicks on the "available" thing for the first time. Possible @HBelusca ?
> > 
> 
> That is one of the options for rapps already, so just configure that?

Stas'M:
Joachim suggestion is about the RAPPS first run, when there is no database downloaded.

File: base/applications/rapps/include/appdb.h (hunk 2 above)
Comment (katahiromz) [reply base/applications/rapps/include/appdb.h:21]: We can use `GlobalAtom` or `LocalAtom`... Not recommended though...

Comment (learn-more) [reply base/applications/rapps/include/appdb.h:21]: I'd propose to put this in the `SETTINGS_INFO` struct.
This can later (in another PR) be refactored to read all settings from the command line (or a file) as override on the registry settings.
e.g. First read the registry, then read from the command line (as override).

Comment (HBelusca) [reply base/applications/rapps/include/appdb.h]: @learn-more This should now be addressed.

File: base/applications/rapps/gui.cpp [hunk 3]
```diff
@@ -96,7 +96,7 @@ CMainWindow::InitCategoriesList()
 {
     HTREEITEM hRootItemInstalled, hRootItemAvailable;
```
Comment (learn-more) [inline base/applications/rapps/gui.cpp:97]: nit: Could have removed `hRootItemInstalled` entirely.

File: base/applications/rapps/include/gui.h [hunk 4]
```diff
@@ -57,12 +57,14 @@ class CMainWindow : public CWindowImpl<CMainWindow, CWindow, CFrameWinTraits>
...
     BOOL bUpdating = FALSE;
+    BOOL bAppwizMode;
+    HTREEITEM hTreeInstalled;
 
     CStringW szSearchPattern;
     AppsCategories SelectedEnumType;
 
   public:
-    CMainWindow(CAppDB *db);
+    CMainWindow(CAppDB *db, BOOL bAppwiz = FALSE);
```
Comment (learn-more) [inline base/applications/rapps/include/gui.h:67]: nit: I am not a fan of default arguments for constructors.

Comment (learn-more) [reply base/applications/rapps/include/gui.h:67]: If you decide to keep it, then this ctor should be marked `explicit` (yes, that needed to be there before as well)

File: base/applications/rapps/unattended.cpp [hunk 5]
```diff
@@ -253,10 +259,17 @@ ParseCmdAndExecute(LPWSTR lpCmdLine, BOOL bIsFirstLaunch, int nCmdShow)
             /* Activate window */
             ShowWindow(hWindow, SW_SHOWNORMAL);
             SetForegroundWindow(hWindow);
+            if (bAppwizMode)
+                PostMessage(hWindow, WM_COMMAND, ID_ACTIVATE_APPWIZ, 0);
             return FALSE;
         }
 
-        MainWindowLoop(&db, nCmdShow);
+        CMainWindow *wnd = new CMainWindow(&db, bAppwizMode);
```
Comment (learn-more) [inline base/applications/rapps/unattended.cpp:267]: Don't need to manually allocate this anymore, can just create it on the stack:
```suggestion
        CMainWindow wnd(&db, bAppwizMode);
```

**Review by learn-more — APPROVED** [review]
Looks a lot better now, I have some minor nitpicks but nothing too serious.

File: base/applications/rapps/include/gui.h (hunk 4 above)
Comment (HBelusca) [reply base/applications/rapps/include/gui.h]: Addressed with the `explicit`

**Comment by HBelusca** [comment]
All last-time comments have been addressed, it should be good to go!
